lib/activerecord/all/activerecord.rbi

```diff
@@ -1,36 +1,775 @@
+      class_name: T.nilable(T.any(Symbol, String)),
+      counter_cache: T.nilable(T::Boolean),
+      dependent: T.nilable(T.any(Symbol, String)),
+      foreign_key: T.nilable(T.any(Symbol, String)),
+      foreign_type: T.nilable(T.any(Symbol, String)),
+      inverse_of: T.nilable(T.any(Symbol, String)),
+      optional: T.nilable(T::Boolean),
+      polymorphic: T.nilable(T.any(Symbol, String)),
+      primary_key: T.nilable(T.any(Symbol, String)),
+      required: T.nilable(T::Boolean),
+      touch: T.nilable(T::Boolean),
+      validate: T.nilable(T::Boolean),
+      default: T.nilable(T.proc.returns(T.untyped))
+    ).void
+  end
+  def belongs_to(
+    name,
+    scope = nil,
+    autosave: nil,
+    class_name: nil,
+    counter_cache: nil,
+    dependent: nil,
+    foreign_key: nil,
+    foreign_type: nil,
+    inverse_of: nil,
+    optional: nil,
     polymorphic: nil,
+    primary_key: nil,
+    required: nil,
     touch: nil,
     validate: nil,
-    optional: nil
+    default: nil
+  ); end
+
+  sig do
+    params(
+      name: T.nilable(T.any(Symbol, String)),
+      scope: T.nilable(T.proc.void),
+      after_add: T.nilable(T.any(Symbol, String, T.proc.void)),
+      after_remove: T.nilable(T.any(Symbol, String, T.proc.void)),
+      association_foreign_key: T.nilable(T.any(Symbol, String)),
+      autosave: T.nilable(T::Boolean),
+      before_add: T.nilable(T.any(Symbol, String, T.proc.void)),
+      before_remove: T.nilable(T.any(Symbol, String, T.proc.void)),
+      blk: T.nilable(T.proc.void),
+      class_name: T.nilable(T.any(Symbol, String)),
+      extend: T.nilable(T.any(Module, T::Array[Module])),
+      foreign_key: T.nilable(T.any(Symbol, String)),
+      join_table: T.nilable(T.any(Symbol, String)),
+      validate: T.nilable(T::Boolean),
+    ).void
+  end
+  def has_and_belongs_to_many(
+    name,
+    scope = nil,
+    after_add: nil,
+    after_remove: nil,
+    association_foreign_key: nil,
+    autosave: nil,
+    before_add: nil,
+    before_remove: nil,
+    class_name: nil,
+    extend: nil,
+    foreign_key: nil,
+    join_table: nil,
+    validate: nil,
+    &blk
+  ); end
+end
+
+module ActiveRecord::NestedAttributes::ClassMethods
+  sig do
+    params(
+      attr_names: T.any(T.any(Symbol, String), T::Array[T.any(Symbol, String)]),
+      allow_destroy: T.nilable(T::Boolean),
+      reject_if: T.any(Symbol, T.proc.returns(T::Boolean)),
+      limit: T.any(Integer, Symbol, T.proc.returns(Integer)),
+      update_only: T.nilable(T::Boolean),
+    ).void
+  end
+  def accepts_nested_attributes_for(
+    attr_names,
+    allow_destroy: nil,
+    reject_if: nil,
+    limit: nil,
+    update_only: nil
+  ); end
+end
+
+module ActiveRecord::Scoping::Named::ClassMethods
+  sig do
+    params(
+      name: T.nilable(T.any(Symbol, String)),
+      body: T.nilable(T.any(
+        T.proc.returns(T.untyped),
+        T.proc.params(args: T.nilable(T.untyped)).returns(T.untyped)
```

Comment:
I think this will be interpreted as: either a proc that doesn't accept argument, or a proc that only accept 1 argument
It'd be much nicer to have something like in https://github.com/sorbet/sorbet/issues/1142

Then I'd be able to generate something [like this](https://sorbet.run/#%23%20typed%3A%20true%0A%0Aclass%20District%3A%3ARelation%0A%20%20extend%20T%3A%3ASig%0Aend%0A%0Amodule%20District%3A%3AClassMethods%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig%20%7B%20%0A%20%20%20%20params(%0A%20%20%20%20%20%20name%3A%20Symbol%2C%20%0A%20%20%20%20%20%20body%3A%20T.proc.bind(District%3A%3ARelation).returns(T.untyped)%2C%20%0A%20%20%20%20%20%20block%3A%20T.nilable(T.proc.void)%0A%20%20%20%20).void%0A%20%20%7D%0A%20%20def%20scope(name%2C%20body%2C%20block%3Dnil)%3B%20end%0Aend%0A%0Aclass%20District%0A%20%20extend%20T%3A%3ASig%0A%20%20extend%20District%3A%3AClassMethods%0A%0A%20%20scope%20%3Ano_param%2C%20-%3E%20%7B%0A%20%20%20%20T.reveal_type(self)%0A%20%20%7D%0A%0A%20%20scope%20%3Atwo_params%2C%20-%3E%20(a%2C%20b)%20%7B%0A%20%20%20%20T.reveal_type(self)%0A%20%20%7D%0Aend%0A) that works beautifully

Reply:
Great point! Scope functions can be arbitrarily variadic, right?

Since sorbet doesn't support that, how about this: 

```ruby
VariadicUntypedFunction = T.type_alias(
  T.any(
    T.proc.returns(T.untyped),
    T.proc.params(arg0: T.untyped).returns(T.untyped),
    T.proc.params(arg0: T.untyped, arg1: T.untyped).returns(T.untyped),
    T.proc.params(arg0: T.untyped, arg1: T.untyped, arg2: T.untyped).returns(T.untyped),
    T.proc.params(arg0: T.untyped, arg1: T.untyped, arg2: T.untyped, arg3: T.untyped).returns(T.untyped),
    T.proc.params(arg0: T.untyped, arg1: T.untyped, arg2: T.untyped, arg3: T.untyped, arg4: T.untyped).returns(T.untyped),
    T.proc.params(arg0: T.untyped, arg1: T.untyped, arg2: T.untyped, arg3: T.untyped, arg4: T.untyped, arg5: T.untyped).returns(T.untyped),
    T.proc.params(arg0: T.untyped, arg1: T.untyped, arg2: T.untyped, arg3: T.untyped, arg4: T.untyped, arg5: T.untyped, arg6: T.untyped).returns(T.untyped),
    T.proc.params(arg0: T.untyped, arg1: T.untyped, arg2: T.untyped, arg3: T.untyped, arg4: T.untyped, arg5: T.untyped, arg6: T.untyped, arg7: T.untyped).returns(T.untyped),
    T.proc.params(arg0: T.untyped, arg1: T.untyped, arg2: T.untyped, arg3: T.untyped, arg4: T.untyped, arg5: T.untyped, arg6: T.untyped, arg7: T.untyped, arg8: T.untyped).returns(T.untyped),
    T.proc.params(arg0: T.untyped, arg1: T.untyped, arg2: T.untyped, arg3: T.untyped, arg4: T.untyped, arg5: T.untyped, arg6: T.untyped, arg7: T.untyped, arg8: T.untyped, arg9: T.untyped).returns(T.untyped)
    # Currently Sorbet is limited to procs with 9 args: https://github.com/sorbet/sorbet/blob/df6085e6ae9846f033064513ea14f069c68b0bf9/core/SymbolRef.h#L387
  )
)

module ActiveRecord::Scoping::Named::ClassMethods
  sig do
    params(
      name: T.nilable(T.any(Symbol, String)),
      body: VariadicUntypedFunction,
      blk: T.nilable(VariadicUntypedFunction)
    ).void
  end
  def scope(
    name,
    body,
    &blk
  ); end
end
```

Reply:
I feel like this should be suggested as a core Sorbet feature, some way to have a type with a varying number of parameters.

Or maybe their intent is to discourage stuff like this.